File: packages/relay-core/python/r2_relay_core/transport.py
from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Awaitable, Callable

from .checkpoint import CheckpointStore, InMemoryCheckpointStore, default_checkpoint_state, has_seen_message, remember_message
from .errors import CASRetryExceededError, PreconditionFailedError
from .keyspace import RelayKeyspace, extract_timestamp_from_key
from .object_store import ObjectStore
from .types import CheckpointState, InboxBatch, JSONDict


logger = logging.getLogger(__name__)


class R2RelayTransport:
    def __init__(
        self,
        store: ObjectStore,
        keyspace: RelayKeyspace,
        peer_id: str,
        checkpoint_store: CheckpointStore | None = None,
        sleep: Callable[[float], Awaitable[None]] = asyncio.sleep,
    ) -> None:
        self.store = store
        self.keyspace = keyspace
        self.peer_id = peer_id
        self.checkpoint_store = checkpoint_store or InMemoryCheckpointStore(default_checkpoint_state())
        self.sleep = sleep
        self._send_lanes: dict[str, asyncio.Lock] = {}
        self._checkpoint_state: CheckpointState | None = None

    async def publish_identity(self, payload: JSONDict | None = None) -> JSONDict:
        identity: JSONDict = {
            "peer": self.peer_id,
            "display_name": self.peer_id,
            "role": "server",
            "last_seen": self._now_ms(),
            "protocol": {"name": "r2-relay", "version": 3},
            "software": {"id": "unknown"},
            "capabilities": {
                "messaging": {"text": True, "streaming": False, "reactions": False, "system_events": False},
                "conversations": {"list": False, "create": False, "reset": False, "archive": False, "threading": False},
                "agents": {"list": False, "multiple": False, "switch": False, "per_agent_models": False},
            },
            "limits": None,
            "agents": [],
            "conversations": [],
        }
        if payload:
            identity.update(payload)
        key = self.keyspace.make_identity_key(str(identity["peer"]))
        await self.store.put_object(key, json.dumps(identity), "application/json")
        return identity

    async def get_head_state(self, peer: str) -> dict[str, Any] | None:
        return await self.store.get_json_with_etag(self.keyspace.make_head_key(peer))

    async def read_message(self, key: str) -> JSONDict | None:
        response = await self.store.get_object(key)
        if response is None:
            return None
        if isinstance(response, dict) and "payload" in response:
            payload = response["payload"]
            if isinstance(payload, dict):
                return dict(payload)
            if isinstance(payload, str):
                parsed = json.loads(payload)
                return parsed if isinstance(parsed, dict) else None
        if isinstance(response, dict) and "Body" in response:
            text = await _body_to_text(response["Body"])
            parsed = json.loads(text)
            return parsed if isinstance(parsed, dict) else None
        if isinstance(response, dict):
            return dict(response)
        if isinstance(response, str):
            parsed = json.loads(response)
            return parsed if isinstance(parsed, dict) else None
        return None

    async def send_message(
        self,
        to: str,
        options: dict[str, Any],
    ) -> dict[str, str]:
        lock = self._send_lanes.setdefault(to, asyncio.Lock())
        async with lock:
            return await self._send_message_unlocked(to, dict(options))

    async def send_streaming_snapshots(
        self,
        to: str,
        snapshots: list[str],
        options: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        resolved_options = dict(options or {})
        route = resolved_options["route"]
        stream_id = resolved_options.get("stream_id") or self.keyspace.short_uuid()
        base_attachments = resolved_options.get("attachments")
        results: list[dict[str, str]] = []
        non_empty = [str(s or "") for s in snapshots if str(s or "").strip()]
        for index, text in enumerate(non_empty):
            content: dict[str, Any] = {"type": "text", "text": text}
            if base_attachments:
                content["attachments"] = base_attachments
            delivery: dict[str, Any] = {
                "stream": {
                    "stream_id": stream_id,
                    "seq": index + 1,
                    "state": "final" if index == len(non_empty) - 1 else "partial",
                }
            }
            results.append(
                await self.send_message(to, {"route": route, "content": content, "delivery": delivery})
            )
        return {"stream_id": stream_id, "results": results}

    async def collect_inbox_messages(self, self_id: str, last_seen_key: str | None) -> InboxBatch:
        head_state = await self.get_head_state(self_id)
        head = head_state.get("body") if isinstance(head_state, dict) else None
        if not head or not head.get("head_key"):
            return {"head": head, "messages": []}
        if head.get("head_key") == last_seen_key:
            return {"head": head, "messages": []}

        current = str(head["head_key"])
        messages: list[dict[str, Any]] = []
        visited: set[str] = set()
        while current and current != last_seen_key:
            if current in visited:
                break
            if len(messages) >= 500:
                break
            visited.add(current)
            message = await self.read_message(current)
            if not message:
                break
            messages.append({"key": current, "message": message})
            current = message.get("prev_key") or None

        messages.reverse()
        return {"head": head, "messages": messages}

    async def mark_message_processed(self, key: str, patch: dict[str, Any] | None = None) -> JSONDict | None:
        message = await self.read_message(key)
        if message is None:
            return None
        resolved_patch = dict(patch or {})
        updated: JSONDict = {
            **message,
            "status": {
                "state": resolved_patch.get("state", "done"),
                "processed_at": resolved_patch.get("processedAt", self._now_ms()),
                "processed_by": resolved_patch.get("processedBy", self.peer_id),
            },
        }
        await self.store.put_object(key, json.dumps(updated), "application/json")
        return updated

    async def sweep_by_key_timestamp(self, prefix: str, ttl_days: int) -> dict[str, Any]:
        cutoff_ms = self._now_ms() - ttl_days * 24 * 60 * 60 * 1000
        continuation_token = None
        scanned = 0
        deleted = 0
        while True:
            page = await self.store.list_prefix_page(prefix, continuation_token=continuation_token, max_keys=1000)
            to_delete: list[str] = []
            for item in page.get("contents") or []:
                key = item.get("Key") if isinstance(item, dict) else None
                if not key:
                    continue
                scanned += 1
                ts = extract_timestamp_from_key(key)
                if ts is not None and ts < cutoff_ms:
                    to_delete.append(key)
            if to_delete:
                await self.store.delete_objects(to_delete)
                deleted += len(to_delete)
            if not page.get("is_truncated") or not page.get("next_continuation_token"):
                break
            continuation_token = page["next_continuation_token"]
        return {"prefix": prefix, "scanned": scanned, "deleted": deleted}

    async def poll_inbox(
        self,
        self_id: str,
        handler: Callable[[JSONDict, str], Awaitable[None]],
        poll_interval_ms: int = 5000,
        backoff_max_ms: int = 40000,
        delete_after_processing: bool = True,
        abort_event: asyncio.Event | None = None,
    ) -> None:
        interval_ms = poll_interval_ms
        state = self._load_checkpoint_state()
        try:
            while not (abort_event and abort_event.is_set()):
                batch = await self.collect_inbox_messages(self_id, state.get("last_head_key"))
                if batch["messages"]:
                    for item in batch["messages"]:
                        msg_id = item["message"].get("msg_id") if isinstance(item.get("message"), dict) else None
                        if has_seen_message(state, object_key=item["key"], msg_id=msg_id):
                            state["last_head_key"] = item["key"]
                            self._save_checkpoint_state(state)
                            continue

                        await handler(item["message"], item["key"])
                        state = remember_message(state, object_key=item["key"], msg_id=msg_id)
                        state["last_head_key"] = item["key"]
                        self._save_checkpoint_state(state)

                        if delete_after_processing:
                            await self.store.delete_object(item["key"])
                    interval_ms = poll_interval_ms
                else:
                    await self.sleep(interval_ms / 1000)
                    if not (batch.get("head") or {}).get("head_key"):
                        interval_ms = min(interval_ms * 2, backoff_max_ms)
                    continue
                await self.sleep(interval_ms / 1000)
        finally:
            self._save_checkpoint_state(state)

    async def _send_message_unlocked(
        self,
        to: str,
        options: dict[str, Any],
    ) -> dict[str, str]:
        now_ms = self._now_ms()
        head_key = self.keyspace.make_head_key(to)
        route = options["route"]
        content = options["content"]
        delivery = options.get("delivery")
        for _attempt in range(1, 9):
            current_head = await self.get_head_state(to)
            current_head_body = current_head.get("body") if isinstance(current_head, dict) else None
            prev_key = current_head_body.get("head_key") if isinstance(current_head_body, dict) else None
            key = self.keyspace.make_msg_key(to, now_ms=now_ms)
            message_id = self.keyspace.short_uuid()
            text = content.get("text", "") if isinstance(content, dict) else ""
            message: JSONDict = {
                "msg_id": message_id,
                "from": self.peer_id,
                "to": to,
                "ts_sent": now_ms,
                "prev_key": prev_key,
                "route": route,
                "content": content,
                "delivery": delivery,
                "status": None,
                "size": len(text.encode("utf-8")) if content.get("type") == "text" else None,
            }
            new_head = {
                "head_key": key,
                "head_msg_id": message_id,
                "head_ts": now_ms,
            }
            await self.store.put_object(key, json.dumps(message), "application/json", if_none_match="*")
            try:
                if current_head and current_head.get("etag"):
                    await self.store.put_object(
                        head_key,
                        json.dumps(new_head),
                        "application/json",
                        if_match=current_head["etag"],
                    )
                else:
                    await self.store.put_object(
                        head_key,
                        json.dumps(new_head),
                        "application/json",
                        if_none_match="*",
                    )
                return {"key": key, "message_id": message_id}
            except PreconditionFailedError:
                await self.sleep(0.02)
        raise CASRetryExceededError("Failed to append message after CAS retries")

    def _load_checkpoint_state(self) -> CheckpointState:
        if self._checkpoint_state is None:
            self._checkpoint_state = self.checkpoint_store.load()
        return {
            "last_head_key": self._checkpoint_state.get("last_head_key"),
            "seen": list(self._checkpoint_state.get("seen") or []),
            "seen_msg_ids": list(self._checkpoint_state.get("seen_msg_ids") or []),
        }

    def _save_checkpoint_state(self, state: CheckpointState) -> None:
        snapshot: CheckpointState = {
            "last_head_key": state.get("last_head_key"),
            "seen": list(state.get("seen") or []),
            "seen_msg_ids": list(state.get("seen_msg_ids") or []),
        }
        self.checkpoint_store.save(snapshot)
        self._checkpoint_state = snapshot

    def _now_ms(self) -> int:
        return self.keyspace.now_ms_factory()


async def _body_to_text(body: Any) -> str:
    chunks = await _body_to_chunks(body)
    return b"".join(chunks).decode("utf-8")


async def _body_to_chunks(body: Any) -> list[bytes]:
    if hasattr(body, "__aiter__"):
        chunks: list[bytes] = []
        async for chunk in body:
            chunks.append(chunk if isinstance(chunk, bytes) else bytes(chunk))
        return chunks
    return await asyncio.to_thread(_sync_body_to_chunks, body)


def _sync_body_to_chunks(body: Any) -> list[bytes]:
    if hasattr(body, "iter_chunks") and callable(body.iter_chunks):
        iterator = body.iter_chunks()
    elif hasattr(body, "read") and callable(body.read):
        data = body.read()
        if data is None:
            return []
        iterator = [data]
    elif hasattr(body, "__iter__"):
        iterator = iter(body)
    else:
        raise TypeError(f"Unsupported response body type: {type(body)!r}")
    chunks: list[bytes] = []
    for chunk in iterator:
        if isinstance(chunk, tuple) and chunk:
            chunk = chunk[0]
        if chunk is None:
            continue
        chunks.append(chunk if isinstance(chunk, bytes) else bytes(chunk))
    return chunks


R2RelayService = R2RelayTransport

File: integrations/openclaw/src/relay-core/transport.ts
import { CASRetryExceededError, PreconditionFailedError } from "./errors.js";
import { defaultCheckpointState, hasSeenMessage, InMemoryCheckpointStore, rememberMessage, type CheckpointStore } from "./checkpoint.js";
import { extractTimestampFromRelayKey, RelayKeyspace } from "./keyspace.js";
import { ObjectStore } from "./object-store.js";
import {
  HeadDoc,
  IdentityDoc,
  InboxBatch,
  InboxMessage,
  RelayMessage,
  RelayMessageStatus,
  RelayTextContent,
  RelayRetentionConfig,
  SendMessageOptions,
  SendMessageResult,
  SweepRuleSummary,
} from "./types.js";

async function readBodyAsText(body: unknown): Promise<string> {
  if (typeof body === "string") {
    return body;
  }
  if (!body) {
    return "";
  }
  const value = body as {
    transformToString?: () => Promise<string>;
    [Symbol.asyncIterator]?: () => AsyncIterator<unknown>;
  };
  if (typeof value.transformToString === "function") {
    return await value.transformToString();
  }
  if (typeof value[Symbol.asyncIterator] === "function") {
    const chunks: Buffer[] = [];
    for await (const chunk of value as AsyncIterable<unknown>) {
      chunks.push(Buffer.from(chunk as Uint8Array));
    }
    return Buffer.concat(chunks).toString("utf8");
  }
  return String(body);
}

export interface RelayTransportOptions {
  store: ObjectStore;
  keyspace?: RelayKeyspace;
  peerId: string;
  checkpointStore?: CheckpointStore;
  sleep?: (ms: number) => Promise<void>;
}

export class R2RelayTransport {
  readonly store: ObjectStore;
  readonly keyspace: RelayKeyspace;
  readonly peerId: string;
  readonly checkpointStore: CheckpointStore;
  readonly sleep: (ms: number) => Promise<void>;
  private readonly sendLanes = new Map<string, Promise<SendMessageResult>>();

  constructor(options: RelayTransportOptions) {
    this.store = options.store;
    this.keyspace = options.keyspace ?? new RelayKeyspace();
    this.peerId = options.peerId;
    this.checkpointStore = options.checkpointStore ?? new InMemoryCheckpointStore(defaultCheckpointState());
    this.sleep = options.sleep ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));
  }

  async publishIdentity(identity?: Partial<IdentityDoc>): Promise<IdentityDoc> {
    const doc: IdentityDoc = {
      display_name: this.peerId,
      role: "server",
      last_seen: this.keyspace.nowMsFactory(),
      protocol: { name: "r2-relay", version: 3 },
      software: { id: "unknown" },
      capabilities: {
        messaging: { text: true, streaming: false, reactions: false, system_events: false },
        conversations: { list: false, create: false, reset: false, archive: false, threading: false },
        agents: { list: false, multiple: false, switch: false, per_agent_models: false },
      },
      limits: null,
      agents: [],
      conversations: [],
      ...(identity ?? {}),
      peer: identity?.peer ?? this.peerId,
    };
    await this.store.putObject(this.keyspace.makeIdentityKey(doc.peer), JSON.stringify(doc), "application/json");
    return doc;
  }

  async getHeadState(peer: string): Promise<{ doc: HeadDoc; etag: string | null } | null> {
    const result = await this.store.getJsonWithEtag(this.keyspace.makeHeadKey(peer));
    return result ? { doc: result.body as HeadDoc, etag: result.etag } : null;
  }

  async getHead(peer: string): Promise<HeadDoc | null> {
    return (await this.getHeadState(peer))?.doc ?? null;
  }

  async readMessage(key: string): Promise<RelayMessage | null> {
    const response = await this.store.getObject(key);
    if (!response) {
      return null;
    }
    if (typeof response === "object" && response !== null && "payload" in response) {
      const payload = (response as { payload: unknown }).payload;
      if (typeof payload === "string") {
        return JSON.parse(payload) as RelayMessage;
      }
      return payload as RelayMessage;
    }
    if (typeof response === "object" && response !== null && "Body" in response) {
      return JSON.parse(await readBodyAsText((response as { Body: unknown }).Body)) as RelayMessage;
    }
    if (typeof response === "string") {
      return JSON.parse(response) as RelayMessage;
    }
    return response as RelayMessage;
  }

  async sendMessage(
    to: string,
    options: SendMessageOptions,
  ): Promise<SendMessageResult> {
    const previous = this.sendLanes.get(to) ?? Promise.resolve({ key: "", messageId: "" });
    const current = previous.catch(() => ({ key: "", messageId: "" })).then(() => this.sendMessageUnlocked(to, options));
    this.sendLanes.set(to, current);
    try {
      return await current;
    } finally {
      if (this.sendLanes.get(to) === current) {
        this.sendLanes.delete(to);
      }
    }
  }

  private async sendMessageUnlocked(
    to: string,
    options: SendMessageOptions,
  ): Promise<SendMessageResult> {
    const now = this.keyspace.nowMsFactory();
    const headKey = this.keyspace.makeHeadKey(to);
    for (let attempt = 0; attempt < 8; attempt += 1) {
      const currentHead = await this.getHeadState(to);
      const prevKey = currentHead?.doc.head_key ?? null;
      const key = this.keyspace.makeMsgKey(to, now);
      const msgId = this.keyspace.shortUuid();
      const msg: RelayMessage = {
        msg_id: msgId,
        from: this.peerId,
        to,
        ts_sent: now,
        prev_key: prevKey,
        route: options.route,
        content: options.content,
        delivery: options.delivery ?? null,
        status: null,
        size: options.content.type === "text"
          ? Buffer.byteLength((options.content as RelayTextContent).text ?? "")
          : null,
      };
      const newHead: HeadDoc = {
        head_key: key,
        head_msg_id: msgId,
        head_ts: now,
      };
      await this.store.putObject(key, JSON.stringify(msg), "application/json", undefined, undefined, "*");
      try {
        if (!currentHead?.doc) {
          await this.store.putObject(headKey, JSON.stringify(newHead), "application/json", undefined, undefined, "*");
        } else {
          if (!currentHead.etag) {
            throw new Error("Missing head ETag for CAS update");
          }
          await this.store.putObject(headKey, JSON.stringify(newHead), "application/json", undefined, currentHead.etag, undefined);
        }
        return { key, messageId: msgId };
      } catch (err: unknown) {
        if (!(err instanceof PreconditionFailedError)) {
          throw err;
        }
        await this.sleep(20);
      }
    }
    throw new CASRetryExceededError();
  }

  async sendStreamingSnapshots(
    to: string,
    snapshots: string[],
    options: { route: SendMessageOptions["route"]; streamId?: string | null; attachments?: RelayTextContent["attachments"] },
  ): Promise<{ streamId: string; results: SendMessageResult[] }> {
    const streamId = options.streamId ?? this.keyspace.shortUuid();
    const results: SendMessageResult[] = [];
    const nonEmpty = snapshots.filter((s) => (s ?? "").trim());
    for (let index = 0; index < nonEmpty.length; index += 1) {
      const text = nonEmpty[index] ?? "";
      const result = await this.sendMessage(to, {
        route: options.route,
        content: {
          type: "text",
          text,
          ...(options.attachments?.length ? { attachments: options.attachments } : {}),
        },
        delivery: {
          stream: {
            stream_id: streamId,
            seq: index + 1,
            state: index === nonEmpty.length - 1 ? "final" : "partial",
          },
        },
      });
      results.push(result);
    }
    return { streamId, results };
  }

  async markMessageProcessed(
    key: string,
    patch: {
      processedAt?: number;
      processedBy?: string | null;
      state?: RelayMessageStatus["state"];
      error?: string | null;
    },
  ): Promise<RelayMessage | null> {
    const msg = await this.readMessage(key);
    if (!msg) {
      return null;
    }
    const updated: RelayMessage = {
      ...msg,
      status: {
        state: patch.state ?? "done",
        processed_at: patch.processedAt ?? this.keyspace.nowMsFactory(),
        processed_by: patch.processedBy ?? this.peerId,
        error: patch.error ?? null,
      },
    };
    await this.store.putObject(key, JSON.stringify(updated), "application/json");
    return updated;
  }

  async collectInboxMessages(selfId: string, lastSeenKey: string | null): Promise<InboxBatch> {
    const head = await this.getHead(selfId);
    if (!head?.head_key || head.head_key === lastSeenKey) {
      return { head, messages: [], checkpointHeadKey: lastSeenKey };
    }
    let current: string | null = head.head_key;
    const window: InboxMessage[] = [];
    const visitedKeys = new Set<string>();
    while (current && current !== lastSeenKey) {
      if (visitedKeys.has(current)) {
        break;
      }
      visitedKeys.add(current);
      const msg = await this.readMessage(current);
      if (!msg) {
        break;
      }
      window.push({ key: current, message: msg });
      if (window.length > 500) {
        window.shift();
      }
      current = msg.prev_key ?? null;
    }
    window.reverse();
    return {
      head,
      messages: window,
      checkpointHeadKey: window[window.length - 1]?.key ?? lastSeenKey,
    };
  }

  async sweepByKeyTimestamp(prefix: string, ttlDays: number, abortSignal?: AbortSignal): Promise<SweepRuleSummary> {
    const cutoffTs = this.keyspace.nowMsFactory() - ttlDays * 24 * 60 * 60 * 1000;
    let continuationToken: string | undefined;
    let scanned = 0;
    let deleted = 0;
    while (!abortSignal?.aborted) {
      const page = await this.store.listPrefixPage(prefix, continuationToken, 1000);
      const toDelete: string[] = [];
      for (const item of page.contents) {
        const key = item.Key;
        if (!key) continue;
        scanned += 1;
        const ts = extractTimestampFromRelayKey(key);
        if (ts !== null && ts < cutoffTs) {
          toDelete.push(key);
        }
      }
      if (toDelete.length > 0) {
        await this.store.deleteObjects(toDelete);
        deleted += toDelete.length;
      }
      if (!page.isTruncated || !page.nextContinuationToken) {
        break;
      }
      continuationToken = page.nextContinuationToken;
    }
    return { prefix, scanned, deleted };
  }

  async sweepByLastModified(prefix: string, ttlDays: number, abortSignal?: AbortSignal): Promise<SweepRuleSummary> {
    const cutoffTs = this.keyspace.nowMsFactory() - ttlDays * 24 * 60 * 60 * 1000;
    let continuationToken: string | undefined;
    let scanned = 0;
    let deleted = 0;
    while (!abortSignal?.aborted) {
      const page = await this.store.listPrefixPage(prefix, continuationToken, 1000);
      const toDelete: string[] = [];
      for (const item of page.contents) {
        const key = item.Key;
        if (!key) continue;
        scanned += 1;
        const lastModified = item.LastModified?.getTime?.() ?? null;
        if (lastModified !== null && lastModified < cutoffTs) {
          toDelete.push(key);
        }
      }
      if (toDelete.length > 0) {
        await this.store.deleteObjects(toDelete);
        deleted += toDelete.length;
      }
      if (!page.isTruncated || !page.nextContinuationToken) {
        break;
      }
      continuationToken = page.nextContinuationToken;
    }
    return { prefix, scanned, deleted };
  }

  async sweepRetention(ttl: RelayRetentionConfig, abortSignal?: AbortSignal): Promise<SweepRuleSummary[]> {
    const summaries: SweepRuleSummary[] = [];
    if ((ttl.msg ?? 0) > 0) {
      summaries.push(await this.sweepByKeyTimestamp("msg/", ttl.msg as number, abortSignal));
    }
    if ((ttl.att ?? 0) > 0) {
      summaries.push(await this.sweepByKeyTimestamp("att/", ttl.att as number, abortSignal));
    }
    if ((ttl.identity ?? 0) > 0) {
      summaries.push(await this.sweepByLastModified("identity/", ttl.identity as number, abortSignal));
    }
    if ((ttl.head ?? 0) > 0) {
      summaries.push(await this.sweepByLastModified("head/", ttl.head as number, abortSignal));
    }
    return summaries;
  }

  async pollInbox(
    selfId: string,
    handler: (msg: RelayMessage, key: string) => Promise<void>,
    pollIntervalMs = 5000,
    backoffMax = 40000,
    deleteAfterProcessing = true,
    abortSignal?: AbortSignal,
  ): Promise<void> {
    let interval = pollIntervalMs;
    let state = await this.checkpointStore.load();
    while (!abortSignal?.aborted) {
      const batch = await this.collectInboxMessages(selfId, state.lastHeadKey);
      if (batch.messages.length > 0) {
        for (const item of batch.messages) {
          if (hasSeenMessage(state, { msgId: item.message.msg_id, objectKey: item.key })) {
            state = { ...state, lastHeadKey: item.key };
            await this.checkpointStore.save(state);
            continue;
          }
          await handler(item.message, item.key);
          state = rememberMessage(state, { msgId: item.message.msg_id, objectKey: item.key, at: item.message.ts_sent });
          state = { ...state, lastHeadKey: item.key, lastPollAt: this.keyspace.nowMsFactory() };
          await this.checkpointStore.save(state);
          if (deleteAfterProcessing) {
            await this.store.deleteObject(item.key);
          }
        }
        interval = pollIntervalMs;
        await this.sleep(interval);
        continue;
      }
      state = { ...state, lastPollAt: this.keyspace.nowMsFactory() };
      await this.checkpointStore.save(state);
      await this.sleep(interval);
      if (!batch.head?.head_key) {
        interval = Math.min(interval * 2, backoffMax);
      }
    }
  }
}
